### ISSUE TITLE
Fitting rectangles to a building footprint

### ISSUE THREAD

**Issue by vvoovv** [prochitecture/blosm#20]
Flat roofs contain air-vents, entrances to the roof, chimneys and other stuff. The problem is how to position them on the flat roof. It's easy to position the roof element within a rectangle. For example, predefined arrangement of the roof elements can be used for each group of the rectangle sizes.

The problem can be reformulated in the following way: how to fit a number of rectangles with the given minimum dimension to a building footprint which can contain one or more holes.

It's easy to fit a rectangle to a quadrangle:

![image](https://user-images.githubusercontent.com/613295/123290403-eb173f00-d519-11eb-84bd-fa33f0f41534.png)

**Comment by vvoovv** [comment]
For an arbitrary convex polygon I propose to find the longest edge and then use a scan line to get the rectangles that fit to the polygon. The scan line starts from the longest edge and is parallel to it.

**Comment by polarkernel** [comment]
Should we maybe try to specify the task more in detail, before we think about solutions? I am not yet sure if I got the goal correctly, so let me try to start this discussion (maybe there will follow more points later):

- Do we speak about _convex_ polygons only or maybe also about _convex regions_ within a polygon, with results similar to the example in the first image [here](https://mathoverflow.net/questions/105837/get-largest-inscribed-rectangle-of-a-concave-polygon)?
- Can a polygon with holes be convex? 
- How should the minimum dimension of the rectangle be specified? Minimum shorter edge length for example?
- When there are possible rectangles that are larger than the minimum dimension, should the largest one (with the largest shorter edge) be selected?
- Will there be restrictions to the aspect ratio of the rectangle?
- The rectangle should be aligned to the longest edge of the polygon.

**Comment by vvoovv** [comment]
Typically roof objects are aligned along a roof side. Each rectangle should be aligned along a roof side, unless the whole building footprint is curved.

**Comment by vvoovv** [comment]
> * Do we speak about _convex_ polygons only or maybe also about _convex regions_ within a polygon, with results similar to the example in the first image [here](https://mathoverflow.net/questions/105837/get-largest-inscribed-rectangle-of-a-concave-polygon)? There is NO requirement to get the largest inscribed rectangle.

Yes the final result should be similar to example in the first image [here](https://mathoverflow.net/questions/105837/get-largest-inscribed-rectangle-of-a-concave-polygon)

Here is a real footprint from OSM. The related building has the flat roof:

![image](https://user-images.githubusercontent.com/613295/123512710-ac0bf980-d691-11eb-9a5d-e50028cbdb7f.png)

> * Can a polygon with holes be convex?

Yes.


> * How should the minimum dimension of the rectangle be specified? Minimum shorter edge length for example?

Yes, the minimum edge length of the polygon.


> * When there are possible rectangles that are larger than the minimum dimension, should the largest one (with the largest shorter edge) be selected?

Yes


> * Will there be restrictions to the aspect ratio of the rectangle?

No.

> * The rectangle should be aligned to the longest edge of the polygon.

That applies only to a convex polygon.

**Comment by vvoovv** [comment]
I added the OSM file _fit_rectangles/chelny.osm_ to the repo _osm_extracts_, that contains a bunch of concave building footprints. All buildings in that area has flat roofs.

**Comment by vvoovv** [comment]
> The rectangle should be aligned to the longest edge of the polygon.

For a footprint like this
![image](https://user-images.githubusercontent.com/613295/123512710-ac0bf980-d691-11eb-9a5d-e50028cbdb7f.png)

each rectangle should be aligned along the "nearest major" edge of the polygon.

**Comment by vvoovv** [comment]
Added complex test cases:

_fit_rectangles/sirius_delta.osm_
![image](https://user-images.githubusercontent.com/613295/123520445-5d725580-d6b9-11eb-9056-081331e5fcf9.png)

_fit_rectangles/sirius_sport_center.osm_
![image](https://user-images.githubusercontent.com/613295/123520472-901c4e00-d6b9-11eb-821b-1d3656d3a411.png)

**Comment by vvoovv** [comment]
The inscribed rectangles should cover as much area of the polygon as possible. Say, at least 50% of the polygon area.

**Comment by polarkernel** [comment]
I did some research on the Internet. Most of the papers I found either end in a PhD thesis or state that the problem is NP-complete. So, should I find a solution, at least I have the chance to increase my scientific degree.  :)

There are three types of problems, where solutions can be found:

### Find the largest inscribed rectangle in a polygon
The only practical application is to crop a rectangle from a set of images stitched to a panormama, which is quite simple. Most others are quite complicated, some of them work iterative. Solutions are provided mainly by mathematicians. However, all of them find only one single inscribed rectangle, which is not our goal.

### Cover a polygon by rectangles
This is not what we want, but maybe some tools or ideas can be recycled. Finding enclosing rectangles for parts of the polygon could be a starting point to cut the polygon in useful parts or to shrink these rectangles to end inside the polygon:

<img src="https://user-images.githubusercontent.com/48180714/123544063-cb6a5b80-d751-11eb-9c59-a3398ed20ab1.png" width="340">

### Rectangle packing
The objective is to determine whether a given set of small rectangles can be placed inside a given polygon, such that no two small rectangles overlap. A lot of algorithms exist for this task for practical applications, but all of them pack the rectangles in another rectangle and not in a general polygon. Additinally, none of the algorithms I found allows rotations of the rectangles.

So I have not yet much until now. I would prefer simple heuristic solutions. However, the last two complex test cases you added really make me a bit of a headache.

**Comment by vvoovv** [comment]
> I would prefer simple heuristic solutions.

Yes, a simple heuristic solution is what we need.

**Comment by vvoovv** [comment]
A quick idea.

* Generate a straight skeleton for the polygon.
* Consider only those edges of the straight skeleton that don't have common vertices with the polygon.
* If the edge length is greater than the minimal length of the rectangle edge, extrude the edge in both directions to create a rectangle.

**Comment by polarkernel** [comment]
> A quick idea.

That may be a very good starting point, I will check it.

**Comment by polarkernel** [comment]
> A quick idea.

Using the straight skeleton seems to be very useful. Mainly the roof ridges give helpful hints. The length of the edges is not important, even a very short edge may lead to a large rectangle. But the distance to the contour (what we used to compute the height of the roof) gives directly one of the possible dimensions of a rectangle. Here a gallery of some first results from _chelny.osm_:

<img src="https://user-images.githubusercontent.com/48180714/123648044-e8be2900-d828-11eb-83d9-a028da7c27de.png" width="250"> <img src="https://user-images.githubusercontent.com/48180714/123648112-f96e9f00-d828-11eb-98df-a311bee7907b.png" width="250"> <img src="https://user-images.githubusercontent.com/48180714/123648182-07242480-d829-11eb-9c37-c1cea0b420db.png" width="250">
<img src="https://user-images.githubusercontent.com/48180714/123648306-20c56c00-d829-11eb-9ef6-6eda56785fea.png" width="250"> <img src="https://user-images.githubusercontent.com/48180714/123648389-320e7880-d829-11eb-99c3-03de201e2702.png" width="250"> <img src="https://user-images.githubusercontent.com/48180714/123648490-46527580-d829-11eb-88e4-e2985eacc842.png" width="250">

I will try now to generate the rectangles from these roof ridge edges. Do you already have an idea of what a minimum dimension of such rectangles might be? Do you already have a data structure in mind to be used to store the rectangles?

**Comment by vvoovv** [comment]
> Do you already have an idea of what a minimum dimension of such rectangles might be?

2 meters.

> Do you already have a data structure in mind to be used to store the rectangles?

A list of Python tuples:
(rectangle width, rectangle height, 2D coordinate of the leftmost bottommost point of the rectangle, 2D unit vector along the rectangle width)

**Comment by polarkernel** [comment]
Using the straight skeleton is very promizing. The solution is not as simple as I first thought, but it stays to be what I call simple. No PHD thesis and no NP-completeness. Here the results for some examples (click on the images to see them larger):

<img src="https://user-images.githubusercontent.com/48180714/123935513-969a1680-d994-11eb-981e-d7b560ab4224.png" width="250"> <img src="https://user-images.githubusercontent.com/48180714/123935636-b16c8b00-d994-11eb-9c28-08a185067022.png" width="250"> <img src="https://user-images.githubusercontent.com/48180714/123935848-e4168380-d994-11eb-8616-3f1c335a1830.png" width="250">
<img src="https://user-images.githubusercontent.com/48180714/123935957-00b2bb80-d995-11eb-84af-ad6d9fa73a42.png" width="250"> <img src="https://user-images.githubusercontent.com/48180714/123936055-1c1dc680-d995-11eb-9945-46c68be446d1.png" width="250"> <img src="https://user-images.githubusercontent.com/48180714/123936203-4079a300-d995-11eb-860d-ca01c57ac831.png" width="250">

Surprisingly, even the complex examples result in something meaningful (there is still a bug in the right example, but I will be able to solve that):

<img src="https://user-images.githubusercontent.com/48180714/123936542-8e8ea680-d995-11eb-8b03-b57ab06d7543.png" width="250"> <img src="https://user-images.githubusercontent.com/48180714/123936594-9e0def80-d995-11eb-88b6-ce31063781a2.png" width="250">

However, you wouldn't yet be able to understand my current code, it looks still too experimental. I would like to transfer it now into a from that is clean, more optimized and that fits into the application intended by you. To do that I need to know more about your intention:

- Should the code be based on the class _Building_, where the resulting list of rectangles could be an additional attribute of this class? The visualization would then be done using a new renderer. This solution would require to make the OSM roof shape _flat_ be accessible somehow.
- Or should the solution be based on a signature similar to _bpypolyskel_, delivering the list of rectangles as return value?

**Comment by vvoovv** [comment]
Very encouraging results.

> * Should the code be based on the class _Building_, where the resulting list of rectangles could be an additional attribute of this class? The visualization would then be done using a new renderer. This solution would require to make the OSM roof shape _flat_ be accessible somehow.
> * Or should the solution be based on a signature similar to _bpypolyskel_, delivering the list of rectangles as return value?

I suggest to organize it as a library. Similar input like _bpypolyskel_. It may be also needed to generate streets (e.g. fitting a bus stop, an entrance to the subway, etc).

**Comment by polarkernel** [comment]
> I suggest to organize it as a library. Similar input like bpypolyskel. It may be also needed to generate streets (e.g. fitting a bus stop, an entrance to the subway, etc).

OK, fine. shall we create a new repository (if yes, could you create it)? Any proposal for a name of this library?

**Comment by polarkernel** [comment]
Forgot to say that the skeleton used is based on bpypolyskel's function _skeletonize()_. Maybe it should be added there?

**Comment by vvoovv** [comment]
I created the repo https://github.com/prochitecture/fit_rectangles

**Comment by vvoovv** [comment]
The path to _bpypolyskel_ can be added in the examples:

```python
import os

def _checkPath():
    path = os.path.join( os.path.dirname(__file__), "bpypolyskel")
    if path in sys.path:
        sys.path.remove(path)
    # make <path> the first one to search for a module
    sys.path.insert(0, path)
_checkPath()
```

**Comment by vvoovv** [comment]
I added a number of test files to _osm_extracts/fit_rectangles_:
* brussels_eu_headquaters.osm
* brussels_eu_headquaters_bool.osm
* moscow_white_house.osm
* moscow_white_house_bool.osm
* washington_pentagon.osm

Bpypolyskel fails to create a straight skeleton for _washington_pentagon.osm_.

**Comment by polarkernel** [comment]
> A list of Python tuples:
(rectangle width, rectangle height, 2D coordinate of the leftmost bottommost point of the rectangle, 2D unit vector along the rectangle width)

Currently, I get the result as four vertices in counter-clockwise order. You do not happen to have an algorithm at your fingertips, that converts these to the desired form? I stuck on finding a conversion.

**Comment by vvoovv** [comment]
> Currently, I get the result as four vertices in counter-clockwise order.

Let's keep it in this form for now.

**Comment by polarkernel** [comment]
> Let's keep it in this form for now.

I commited a version, that works in my experimental environment just as backup. The interface, similar to _polygonize()_ in _bpypolyskel_, is explained in the code.

> The path to bpypolyskel can be added in the examples:

I will try to understand and integrate that tomorrow.

**Comment by polarkernel** [comment]
> The path to bpypolyskel can be added in the examples:

I am sorry, I do not yet understand how you like to get _fit_rectangles_. Currently, it has a similar interface to _polygonize()_ in _bpypolyskel_:

```Python
def fit_rectangles(verts, firstVertIndex, numVerts, holesInfo=None, unitVectors=None)
```
As dependency, it uses _bpypolyskel_ and _mathutils_. So I could add _bpypolyskel_ to the repo and write an example of its usage. This would have the disadvantage, that eventual changes in the repo of _bpypolyskel_ would not be replicated automatically to _fit_rectangle_.

My test environment can't be used directly. I get _buildings_ in _facade_visibility_, call there a function that converts the building data to the form used by _bpypolyskel_ using 3D _mathutils_ vectors, call then _fit_rectangles()_ and plot the result locally. Like this I have a lot of buildings available for tests. But this is very experimental and can't be a final solution.

I do not understand what you mean by calling __checkPath_().

**Comment by vvoovv** [comment]
I'll work on integration today and tomorrow.

**Comment by vvoovv** [comment]
I've converted the library _fit_rectangles_ to a pure 2D one.

**Comment by vvoovv** [comment]
I implemented a renderer for the library _fit_rectangles_ in the repo _blosm_. Holes aren't supported yet.

To visualize the results execute:

```
cd /path/to/blosm
python -m script.command_line --buildings --setupScript setup/mpl_fit_rectangles.py --osmFilepath /path/to/file.osm
```

**Comment by vvoovv** [comment]
The repos _blosm_ and _fit_rectangles_ must be located next to each other.

[Here](https://github.com/prochitecture/blosm/blob/a272abba28126c1d596eaa7a190711cf9f56d1ca/setup/mpl_fit_rectangles.py#L5) is how the function __checkPath_ is used to get the library _fit_rectangles_ available in the repo _blosm_.

**Comment by polarkernel** [comment]
> I implemented a renderer for the library fit_rectangles in the repo blosm. Holes aren't supported yet.

Works fine, thanks! It was not so easy to convince VS Code to accept and debug program code outside its starting repo, but now it works.

> Here is how the function _checkPath is used to get the library fit_rectangles available in the repo blosm.

Great trick!

Do you like to get a description on how _fit_rectangle_ works or is the code with comments already clear enough?

**Comment by vvoovv** [comment]
> Do you like to get a description on how _fit_rectangle_ works or is the code with comments already clear enough?

A description would be very helpful.

**Comment by vvoovv** [comment]
Can the code to create 2D edges (Edge2) for the outer ring and holes be shared with _bpypolyskel_?

**Comment by vvoovv** [comment]
To support holes I need to proceed with the integration of the facade classification and feature detection.

**Comment by polarkernel** [comment]
> Can the code to create 2D edges (Edge2) for the outer ring and holes be shared with bpypolyskel?

Yes, but only if we go back to 3D vertices at the input, because _bpypolyskel_ 's function _polygonize()_ needs their z-values. _fit_recatngles_ needs only _edges2D_ and _edgeContours_, which both are already 2D.

**Comment by vvoovv** [comment]
> Yes, but only if we go back to 3D vertices at the input, because _bpypolyskel_ 's function _polygonize()_ needs their z-values. _fit_recatngles_ needs only _edges2D_ and _edgeContours_, which both are already 2D.

Let's keep it as is for now.

**Comment by polarkernel** [comment]
### Algorithm of _fit_rectangles_

First, a straight skeleton of the building's footprint polygon is constructed using the function _skeletonize()_ of _bpypolyskel_:
<img src="https://user-images.githubusercontent.com/48180714/124468208-1e1ec580-dd99-11eb-8108-f4a98f111c8e.png" width="400">
The function delivers the skeleton's nodes and edges. To fit rectangles, only the inner edges are selected, these are the edges that connect the skeleton nodes but not the nodes of the building's outline. As the straight skeleton can be used to construct hipped roofs, its nodes provide additionally an attribute _height_, which corresponds to half of the distance between the edges whose wavefront have met at the skeleton edge. Flat edges with a small slope (less than MAX_SLOPE) between their nodes, which form the roof's ridges, are selected, because their polygon edges are almost parallel. Redundant nodes with almost straight angles between their edges are then removed. The red edges in the following picture correpsond to the selected ridges in th example:
<img src="https://user-images.githubusercontent.com/48180714/124470810-62f82b80-dd9c-11eb-9397-1d95a4076593.png" width="400">
These ridges are sorted by decreasing _height_ of their lower node, because large heights promise large distances between their corresponding edges and therefore large rectangles. Before the loop for the construction of rectangle starts, a list _vertices_ is filled with all the vertices of the edges of the outline polygon and another list _edges_ with the indices of the edges. Later this list will also contain the edges of the rectangles already constructed. Additionally, an empty list _rectangles_ is prepared to be filled later by the rectangles.
#### Rectangle construction loop
The loop starts with the highest ridge in the sorted ridges list. One of its endpoints (the _anchor_) will be the seed for the rectangle fit. It is important that this seed is not enclosed in one of the rectangles. If both endpoints are in one of the rectangles, the midpoint is checked. If none of these three points is outside the rectangles, this ridge is discarded and the next ridge is selected. Using the unit vector of the ridge, all vertices in _vertices_ are rotated so that the ridge (red line) is placed on the x-axis and points from the anchor (red cross) to the right.
<img src="https://user-images.githubusercontent.com/48180714/124473525-ba4bcb00-dd9f-11eb-946b-3c209a58dcf9.png" width="400">
The distance of the ridge to their edges is estimated from the height of the lower node. It is decreased by a small value MARGIN to avoid fast intersection with the almost parallel lines. Two imaginary lines (dotted green) parallel to the x-axis and offset by this distance are now constructed. The x-values of all vertices (green dots) within these limits  of y-values and also those of the intersections of these green lines with the edges (red dots) get collected in a list. A possible axis-aligned rectangle can now be fitted between the green lines, the largest negative x-value and the smallest positive x-value in the list of x-values:
<img src="https://user-images.githubusercontent.com/48180714/124481709-def87080-dda8-11eb-9528-7a89a99e1965.png" width="400">
To accept this rectangle, its width or height has to be larger than the limit MINDIM. The rectangle is then rotated back to the original position and checked for collision with the already existing rectangles. When accepted, it is inserted into the list _rectangles_ and also in the lists _vertices_ and _egdes_. Then the loop continues with the next ridge until these are exhausted. The new rectangle appears then in the rotated edges and may produce intersections:
<img src="https://user-images.githubusercontent.com/48180714/124476639-5b885080-dda3-11eb-9166-2e01c7dc10db.png" width="400">

**Comment by vvoovv** [comment]
Would it be possible to get rid of _numpy_ in the code of _fit_rectangles_?

**Comment by polarkernel** [comment]
> Would it be possible to get rid of numpy in the code of fit_rectangles?

Yes. I will adapt it tomorrow.

**Comment by polarkernel** [comment]
> Would it be possible to get rid of numpy in the code of fit_rectangles?

I committed the version without numpy.

**Comment by vvoovv** [comment]
What does it fit a rectangle to a [rectangle](https://github.com/prochitecture/osm_extracts/blob/master/fit_rectangles/rectangle_01.osm) but fail to fit a rectangle to a [quadrangle](https://github.com/prochitecture/osm_extracts/blob/master/fit_rectangles/quadrangle_01.osm)?

![image](https://user-images.githubusercontent.com/613295/125274244-72d0bc00-e316-11eb-8800-a566f81337aa.png)
![image](https://user-images.githubusercontent.com/613295/125274477-aca1c280-e316-11eb-823d-4ea97b6d1490.png)

**Comment by vvoovv** [comment]
Can a better coverage with rectangles be achieved for this [footprint](https://github.com/prochitecture/osm_extracts/blob/master/fit_rectangles/sirius_bolshoi.osm)?

![image](https://user-images.githubusercontent.com/613295/125276317-c3491900-e318-11eb-8fba-10d17743b107.png)

**Comment by vvoovv** [comment]
Another my concern is what to do with a circular [footprint](https://github.com/prochitecture/osm_extracts/blob/master/fit_rectangles/sirius_planetarium.osm). The straight skeleton for it doesn't have an edge without common vertices with the footprint:
![image](https://user-images.githubusercontent.com/613295/125278054-e674c800-e31a-11eb-9f7c-0835654c5b94.png)

**Comment by polarkernel** [comment]
> What does it fit a rectangle to a rectangle but fail to fit a rectangle to a quadrangle?

Due to slope limit (now removed)

> Can a better coverage with rectangles be achieved for this footprint?

Was a bug in removal of straight angles, now fixed.

> Another my concern is what to do with a circular footprint.

Fixed as special case.

In the currently commited version, I removed the slope limit, because in many cases it was not helpful or even sometimes destructive. Seems still to work, as far as I experienced.

**Comment by vvoovv** [comment]
The coverage with rectangles was better for _osm_extracts/fit_rectangles/brussels_eu_headquaters.osm_ before the latest commit:
Now:
![image](https://user-images.githubusercontent.com/613295/125326672-f86e5f00-e34a-11eb-91b8-fba2be15e8cb.png)
Before:
![image](https://user-images.githubusercontent.com/613295/125326733-0623e480-e34b-11eb-9ed7-a0cca4ec9cc6.png)

**Comment by vvoovv** [comment]
To go to the previous state, checkout to the previous commit:
```
git checkout 78750736dcdcf89e277514db040ba26fe4a1e05c
```

To go to the latest commit:
```
git checkout master
```

**Comment by vvoovv** [comment]
> The coverage with rectangles was better for _osm_extracts/fit_rectangles/brussels_eu_headquaters.osm_ before the latest commit

The same applies for: _osm_extracts/fit_rectangles/brussels_eu_headquaters_bool.osm_

**Comment by vvoovv** [comment]
Also for _osm_extracts/fit_rectangles/chelny.osm_:

![image](https://user-images.githubusercontent.com/613295/125328686-42584480-e34d-11eb-9dd1-e896290cce5d.png)


Now:
![image](https://user-images.githubusercontent.com/613295/125328394-eee5f680-e34c-11eb-8dbc-536b2e3b4eae.png)

Before:
![image](https://user-images.githubusercontent.com/613295/125328433-f86f5e80-e34c-11eb-8485-c8ef91b47c80.png)

**Comment by polarkernel** [comment]
OK, I'm convinced. I reversed partly the last commit. Forcing a minimal slope leads to nearly parallel edges, which is favourable for many buildings. I left the fix for the straight angle removal and the special case for the circle in the code. _quadrangle_01.osm_ will not work, because its single edge without common vertices with the footprint is too steep for the slope limit. The same is valid for many of the skeleton edges of _sirius_bolshoi.osm_, only a short edge remains which leads to a small rectangle.

**Comment by vvoovv** [comment]
Why is a large rectangular area uncovered with rectangles for the attached [footprint](https://github.com/prochitecture/blosm/files/6807287/chelny_01.zip)

![image](https://user-images.githubusercontent.com/613295/125419773-90017e89-7058-48a6-8f28-c29aaaa472b7.png)

**Comment by polarkernel** [comment]
I like to propose an improvement that I am currently testing. The first step of the rectangle construction loop, that I mentioned in my [algorithm description](https://github.com/prochitecture/blosm/issues/20#issuecomment-874132866), consists of two imaginary lines (dotted green in the image below) parallel to the x-axis and offseted by the distance given by the height of the skeleton ridge. The intersections of these lines with the polygon most often determines the size of the rectangle:

<img src="https://user-images.githubusercontent.com/48180714/124473525-ba4bcb00-dd9f-11eb-946b-3c209a58dcf9.png" width="400">

For this example (and mainly for rounded building footprints) the rectangle could be larger when we decrease this distance a bit. In my proposition, this distance starts with the height as currently, but then gets decreased step by step, as long as the area of the possible rectangle increases. For rectilinear buildings, this process stops almost immediately, but for rounded footprints, a much larger rectangle may be found by this scanning. Here some examples extracted from the dataset _sirius.osm_ (left: before, right: this proposition):

<img src="https://user-images.githubusercontent.com/48180714/125425720-94064526-1610-4491-a259-158ef8afe6eb.png" width="400"> <img src="https://user-images.githubusercontent.com/48180714/125425851-67704e32-3c6e-4ebc-bae9-46ddde524909.png" width="400">

The direction of the rectangle still depends on the direction of the skeletons seed edge, as can be seen for the rightmost building. The computation time is naturally increased a bit for the round buildings, but only slightly for the rectilinear ones. Because finding intersections with axis-parallel lines is extremely fast, this kind of scanning is very efficient. Shall I commit this solution?

**Comment by vvoovv** [comment]
> I like to propose an improvement that I am currently testing. 

Does it also solve the above [problem](https://github.com/prochitecture/blosm/issues/20#issuecomment-878898307)?

**Comment by polarkernel** [comment]
> Does it also solve the above problem?

No. This seems to have another reason, but I was not yet able to locate it.

**Comment by polarkernel** [comment]
> Does it also solve the above problem?

Fixed now. Was a problem of a wrong result of rectangle intersections due to floating point tolerances. I had to introduce small margins around rectangles.

> I like to propose an improvement that I am currently testing.

I committed this version now, together with the above fix. The only remaining issue is the case of quadrangles. I will try to solve them as a special case, similar to circular footprints.

**Comment by vvoovv** [comment]
> The only remaining issue is the case of quadrangles. I will try to solve them as a special case, similar to circular footprints.

How about using the approach from the first [message](https://github.com/prochitecture/blosm/issues/20#issue-929347945)?

**Comment by polarkernel** [comment]
Processing of quadrilaterals is implemented. All rectangles are based on the longest edge of the quadrilateral. Here a few examples from _sirius.osm_:

<img src="https://user-images.githubusercontent.com/48180714/125633565-02d311f2-cc3b-4085-b144-25dc160355fb.png" width="200"> <img src="https://user-images.githubusercontent.com/48180714/125633663-16f2e43f-adfd-4ce6-b9b4-e50fc0ed9673.png" width="200"> <img src="https://user-images.githubusercontent.com/48180714/125633763-adfef9a6-9a19-4de6-b6a5-0239eb8f6b4b.png" width="200"> <img src="https://user-images.githubusercontent.com/48180714/125633876-b482e199-7a72-46dc-93c1-780f0b07565e.png" width="200">

**Comment by vvoovv** [comment]
I discovered that _sirius.osm_ contains a number of circular buildings that the library fails to cover with rectangles. Those buildings are [attached](https://github.com/prochitecture/blosm/files/6817087/circular.zip). Is it because those circles aren't perfect?

**Comment by vvoovv** [comment]
I discovered that a number of buildings in _sirius.osm_ doesn't have rectangles. The buildings are
[atttached](https://github.com/prochitecture/blosm/files/6817320/failed.zip). I am not sure if anything can be done with that.

**Comment by polarkernel** [comment]
> I discovered that sirius.osm contains a number of circular buildings that the library fails to cover with rectangles. Those buildings are attached. Is it because those circles aren't perfect?

Yes. The skeleton looks for example like this:

<img src="https://user-images.githubusercontent.com/48180714/125650976-3a946cf2-65b4-415f-aa91-c50f7f9acbb4.png" width="300">

Before I just checked for a single skeleton node and added there a temporary skeleton edge, which worked for perfect circles. Now I replaced it by a temporary skeleton edge, placed at the center of gravity of the existing nodes, if any. This may even save other shapes with weird skeletons.

**Comment by vvoovv** [comment]
> I discovered that a number of buildings in _sirius.osm_ doesn't have rectangles. The buildings are
> [atttached](https://github.com/prochitecture/blosm/files/6817320/failed.zip). I am not sure if anything can be done with that.

The latest commit also solved this issue.

**Comment by vvoovv** [comment]
Is the polygon simplification helpful in the case of concave features? I am afraid it will be faster to keep the concave features as is.

![image](https://user-images.githubusercontent.com/613295/129911051-b122415f-ab6a-4108-a319-4e9ca9932c8b.png)
